cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -1066,20 +1098,88 @@ class ObjectAppendStream final : public io::OutputStream {
       // flush. This also avoids some unhandled errors when flushing in the destructor.
       return Status::OK();
     }
-    return CommitBlockList(block_blob_client_, block_ids_, commit_block_list_options_);
+
+    auto fut = FlushAsync();
+    RETURN_NOT_OK(fut.status());
+
+    return CommitBlockList();
+  }
+
+  Future<> FlushAsync() {
+    RETURN_NOT_OK(CheckClosed("flush"));
```

Comment:
Can we move https://github.com/apache/arrow/blob/0bae073d1abe439d113cc12e06e7ada886a2f2fd/cpp/src/arrow/filesystem/azurefs.cc#L1063-L1067 from `Flush()`?
It seems that we don't need to execute `CheckClosed("flush")` and `if (!initialized_)` in both of `Flush()` and `FlushAsync()`.

Reply:
I moved this to `Flush`, since this is the public API.

Reply:
I think I agree with @kou. The S3 filesystem implements `Flush()` as
```
  Status Flush() override {
    auto fut = FlushAsync();
    return fut.status();
  }
```
I think it would be nice if we could do the same because it ensures that `Flush` and `FlushAsync` have the same behaviour.

Reply:
I tried to unify the sync/async flush implementations, unfortunately that did not work out due to lifetime issues in the async case. When an `ObjectAppendStream` is deconstructed in RAII way, `Close()` (and therefore `Flush()`) is called. If we call `FlushAsync()` in the close, we need to create a `shared_ptr` from `this` (to ensure lifetime of `this` when the lambda is actually called), but: we can not create a `shared_ptr` of `this` while it is deconstructed. Hence in the `Close()` call we always must do a sync `Flush` where we do not have to give these lifetime guarantees. 
TLDR: `Flush()` and `FlushAsync()` impls are similar, but slightly different and decoupled implementations.